feat: [N8N-4365] Google Cloud Storage Node

freyamade opened the pull request.

Commit:
:sparkles: - Credentials file for Google Cloud Storage OAuth

Commit:
:sparkles: - Creating GoogleCloudStorage node

Currently has Bucket.getAll and Bucket.get implemented

Commit:
:shirt: - Fixing lint issues in Bucket.getAll

Commit:
:sparkles: - Added handling for optional query parameters for getAll and get

Commit:
:sparkles: - Added Bucket.create method

Commit:
:sparkles: - Added Bucket.update

Commit:
:sparkles: - Bucket.delete method added

krynble commented on packages/nodes-base/nodes/Google/CloudStorage/BucketDescription.ts:
```diff
@@ -0,0 +1,511 @@
+		default: '',
+	},
+	{
+		displayName: 'Projection',
```
I don't think this property is doing anything. Could you confirm?

freyamade replied:
This parameter is an optional one that basically says whether the response should contain access control data or not basically!

https://cloud.google.com/storage/docs/json_api/v1/buckets/list can see it here for example

freyamade commented on packages/nodes-base/nodes/Google/CloudStorage/BucketDescription.ts:
```diff
@@ -0,0 +1,511 @@
+
+export const bucketFields: INodeProperties[] = [
+	{
+		displayName: 'Project',
```
Sure thing!

Commit:
:art: - Rename project parameter to projectId

Commit:
:bug: - Projection parameter was not being sent

Commit:
:art: - Remove unused imports from Bucket file

Commit:
:bug: - Fixed issue with pagination of empty list

Commit:
:art: - Remove unnecessary interface

Commit:
:sparkle: - Added Return All toggle to Bucket.list

Commit:
:sparkles: - Object.getAll

Commit:
:shirt: - Fixing unbraced if statements

Commit:
:sparkles: - Object.get method

Please note the media download doesn't currently work but I don't want to lose
other dev progress through stashing

Commit:
:sparkles: - Object.delete

Commit:
:shirt: - Fixing lint issues

Commit:
:sparkles: - Fixing JSON issues & Object.update

krynble committed:
Merge branch 'master' into n8n-4365-google-cloud-storage-node-development

krynble committed:
Merge branch 'master' into n8n-4365-google-cloud-storage-node-development

krynble committed:
Merge branch 'master' of github.com:n8n-io/n8n into n8n-4365-google-cloud-storage-node-development

krynble committed:
fix: download files operation correctly setting encoding type

Commit:
:sparkles: - Object.read finally works!

Commit:
:art: - Using PATCH instead of PUT for update requests

Commit:
:sparkles: - Object.create without binary property

Commit:
:sparkles: - Object.create done

krynble commented on packages/nodes-base/nodes/Google/CloudStorage/BucketDescription.ts:
```diff
@@ -0,0 +1,617 @@
+		},
+		options: [
+			{
+				displayName: 'Predefined ACL',
```
We can add a `routing` block to this option, allowing us to effectively remove the `preSend` blocks above from create and update operations.
It should look like
```
				routing: {
					request: {
						qs: {
							predefinedAcl: '={{$value}}',
						}
					}
				},
```

freyamade replied:
Oh yeah good point! The preSend functions for create / update were bigger but because I took out the common JSON body parsing I didn't actually realise that it had simplified down to this, thank you!

freyamade replied:
Oh but I just realised, the `createAcl` is a collection so it doesn't guarantee having either `predefinedAcl` or `predefinedObjectAcl` defined for any request so shouldn't it stay in this function? :thinking:

freyamade replied:
Oh wait I just got it never mind

krynble commented on packages/nodes-base/nodes/Google/CloudStorage/BucketDescription.ts:
```diff
@@ -0,0 +1,617 @@
+						},
+						returnFullResponse: true,
+					},
+					send: {
+						preSend: [
+							async function (this, requestOptions) {
+								if (!requestOptions.qs) requestOptions.qs = {};
+								const additionalQs = this.getNodeParameter('createAcl') as IDataObject;
+
+								// Merge in the options into the queryset
+								requestOptions.qs = Object.assign(requestOptions.qs, additionalQs);
+								return requestOptions;
+							},
+						],
+					},
```
This whole block can be removed with the addition of the routing suggested below.

krynble commented on packages/nodes-base/nodes/Google/CloudStorage/BucketDescription.ts:
```diff
@@ -0,0 +1,617 @@
+							projection: '={{$parameter["projection"]}}',
+						},
+					},
+					send: {
+						preSend: [
+							async function (this, requestOptions) {
+								if (!requestOptions.qs) requestOptions.qs = {};
+								const options = this.getNodeParameter('getFilters') as IDataObject;
+
+								// Merge in the options into the queryset
+								requestOptions.qs = Object.assign(requestOptions.qs, options);
+								return requestOptions;
+							},
+						],
+					},
```
We should be able to remove this block as well.

krynble commented on packages/nodes-base/nodes/Google/CloudStorage/BucketDescription.ts:
```diff
@@ -0,0 +1,617 @@
+
+export const bucketFields: INodeProperties[] = [
+	{
+		displayName: 'Project ID',
+		name: 'projectId',
+		type: 'string',
+		required: true,
+		placeholder: 'Project ID',
+		displayOptions: {
```
I've noticed this argument is sent on every request as query string with the same name, so we can simplify this:
```suggestion
		displayName: 'Project ID',
		name: 'projectId',
		type: 'string',
		required: true,
		placeholder: 'Project ID',
				routing: {
			request: {
				qs: {
					project: '={{ $value }}',
				},
			},
		},
		displayOptions: {
```

krynble reviewed (changes requested):
Amazing job on the node @freyamade 

I esp loved the creativity in using the `do...while` for paginating!

My comments are mostly related to avoid code duplication in a few parts - it makes sense to concentrate the payload, query string and headers as part of the attributes whenever possible.

I noticed that the bucket name is sometimes sent as part of the body or as part of the URL, so in this case it doesn't make sense to centralize it, but I believe many of the `preSend` functions can be removed by following the examples I provided.

Could you please take a quick look and make those small adjustments?

Thanks!

freyamade commented on packages/nodes-base/nodes/Google/CloudStorage/BucketDescription.ts:
```diff
@@ -0,0 +1,617 @@
+};
+
+// Define a JSON parse function here to use it in two places
+async function parseJSONBody(this: IExecuteSingleFunctions, requestOptions: IHttpRequestOptions): Promise<IHttpRequestOptions> {
```
I missed this, it was meant to be used in the Create and Update preSend blocks but I must have forgotten to reference it, thank you!

Commit:
:art: - Moving QS routing to fields as requested

Commit:
Merge branch 'master' of github.com:n8n-io/n8n into n8n-4365-google-cloud-storage-node-development

Commit:
:shirt: - Fixing lint issues

krynble reviewed (approved):
LGTM!

janober commented:
Got released with `n8n@0.194.0`
